_posts/2020-04-03-what-does-appsync-do.md

```diff
@@ -0,0 +1,200 @@
+
+In the schema, we are creating a mutation for create order. This expects the input argument to match the type defined in orderInput. The result returned matches the type defined in Order.
+
+### Step 3: Create the Datasource
+
+The type of data sources supported by AppSync are
+
+- AMAZON_DYNAMODB,
+- AMAZON_ELASTICSEARCH
+- AWS_LAMBDA, HTTP
+- RELATIONAL_DATABASE
+
+In this case, we will use DynamoDb as our data source.
+
+```yml
+CoffeeShopDynamoDB:
+  Type: AWS::DynamoDB::Table
+  Properties:
+    TableName: "CoffeeShopTable"
+    AttributeDefinitions:
+      - AttributeName: "id"
+        AttributeType: "S"
+      - AttributeName: "resourceId"
+        AttributeType: "S"
+    KeySchema:
+      - AttributeName: "id"
+        KeyType: "HASH"
+      - AttributeName: "resourceId"
+        KeyType: "RANGE"
+    ProvisionedThroughput:
+      ReadCapacityUnits: 3
+      WriteCapacityUnits: 1
+    Projection:
+      ProjectionType: "ALL"
+
+CoffeeShopDynamoDBDataSource:
+  Type: AWS::AppSync::DataSource
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    Name: CoffeeShopDynamoDBDataSource
+    Description: Contain order data from coffee shop
+    Type: “AMAZON_DYNAMODB”
+    ServiceRoleArn: arn:aws:iam::xxxxxxxxxxxx:role/xxxxxxxxx
+    DynamoDBConfig:
+      TableName: CoffeeShopTable
+      AwsRegion: ap-southeash-2
+```
+
+From the template, we are provisioning new dynamoDB table for the API. It has the attribute id as the hash key and resourceId as the range key for the table.
+
+We are connecting the AppSync data source with the newly created dynamoDB table by specifying the table name. You can connect to an existing data source if you already have one.
+
+### Step 4: Create the Resolver
+
+Adding resolver for a mutation, query or subscription automatically coverts arguments that match the schema into database operation.
+
+```yml
+CreateOrderResolver:
+  Type: AWS:: AppSync::Resolver
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    TypeName: Mutation
+    FieldName: createOrder
+    DataSourceName: !GetAtt CoffeeShopDynamoDBDataSource.Name
+    RequestMappingTemplate: |
+      {
+        "version" : "2017-02-28",
+        "operation" : "PutItem",
+        "key" : {
+          "id": { "S" : "${ctx.args.orderInput.id}" },
+          "resourceId": { "S": "COFFEE" }
+        },
+      }
+    ResponseMappingTemplate: |
+        $util.toJson($context.result)
+
+GetOrderResolver:
+  Type: AWS:: AppSync::Resolver
+  Properties:
+    ApiId: !GetAtt CoffeeShopAppSyncApi.ApiId
+    TypeName: Query
+    FieldName: getOrder
+    DataSourceName: !GetAtt CoffeeShopDynamoDBDataSource.Name
+    RequestMappingTemplate: |
+      {
+          "version": "2017-02-28",
+          "operation": "GetItem",
+          "key": {
+            "id": { "S": "${ctx.args.id}" },
+            "resourceId": { "S": "COFFEE" }
+          }
+        }
+    ResponseMappingTemplate: |
+        $util.toJson($context.result)
+```
+
```

Comment:
Maybe include a way of testing it? Like what is the call going to look like from the client point of view?

I think it might be beneficial to show the full flow from in terms of:

mutation to create an order -> see the order in the DB (screenshots are fine!) -> get the order back using the query